megamek/src/megamek/server/Server.java

```diff
@@ -3135,71 +3135,78 @@ private void receiveForwardIni(int connectionId) {
+                    return;
+                }
+    
+                // if it is an EntityClassTurn we have to check make sure, that the
+                // turn it is exchanged with is the same kind of turn!
+                // in fact this requires an access function to the mask of an
+                // EntityClassTurn.
+                boolean isEntityClassTurn = (turn instanceof GameTurn.EntityClassTurn);
+                int classMask = 0;
+                if (isEntityClassTurn) {
+                    classMask = ((GameTurn.EntityClassTurn) turn).getTurnCode();
+                }
+    
+                boolean switched = false;
+                int nextTurnId = 0;
+                for (int i = currentTurnIndex; i < turns.size(); i++) {
+                    // if we find a turn for the specific player, swap the current
+                    // player with the player noted there
+                    // and stop
+                    if (turns.get(i).isValid(next.getId(), game)) {
+                        nextTurnId = i;
+                        if (isEntityClassTurn) {
+                            // if we had an EntityClassTurn
+                            if ((turns.get(i) instanceof GameTurn.EntityClassTurn)) {
+                                // and found another EntityClassTurn
+                                if (!(((GameTurn.EntityClassTurn) turns.get(i)).getTurnCode() == classMask)) {
+                                    // both have to refer to the SAME class(es) or
+                                    // they need to be rejected.
+                                    continue;
+                                }
+                            } else {
                                 continue;
                             }
-                        } else {
-                            continue;
                         }
+                        switched = true;
+                        break;
                     }
-                    switched = true;
-                    break;
                 }
+    
+                // update turn order
+                if (switched) {
+                    game.swapTurnOrder(currentTurnIndex, nextTurnId);
+                    // update the turn packages for all players.
+                    send(createTurnVectorPacket());
+                    send(createTurnIndexPacket(connectionId));
+                    return;
+                }
+                // if nothing changed return without doing anything
             }
-
-            // update turn order
-            if (switched) {
-                game.swapTurnOrder(currentTurnIndex, nextTurnId);
-                // update the turn packages for all players.
-                send(createTurnVectorPacket());
-                send(createTurnIndexPacket(connectionId));
-            }
-            // if nothing changed return without doing anything
+            
+            next = game.getTeamForPlayer(current).getNextValidPlayer(next, game);
```

Comment:
Change is here, and can theoretically null error if the Player doesn't have a team assigned (according to a note in getTeamForPlayer, it can return null).